build-system/compile/json-schema/index.js

```diff
@@ -182,27 +182,254 @@ function transformAjvCode(code, scope, config) {
+          continue;
+        }
+      }
+      // Objects may in some cases be referenced only for their key list, in
+      // which we define each unset value as a zero rather than the original.
+      // If the original value is needed, it will be included during a
+      // different loop pass.
+      if (memberExpression && parentIsHasOwnPropertyCall(memberExpression)) {
+        if (typeof replacementAt[key] !== 'object') {
+          replacementAt[key] = {};
+        }
+        // Can no longer consider this level a container, since we require
+        // its keys
+        containers.set(replacementAt[key], false);
+        for (const j in originalAt[key]) {
+          if (!(j in replacementAt[key])) {
+            replacementAt[key][j] = 0;
+          }
+        }
+      } else {
+        replacementAt[key] = originalAt[key];
+      }
+    }
+    /**
+     * Simplify containers so that the full schema is as shallow as possible.
+     * We update container's references to match these changes.
+     * 1. If the container is an object, convert it to an array.
+     * 2. If the container only has one item, replace the container with the
+     *    item.
+     * @param {any} value
+     * @return {any}
+     */
+    function simplifyContainers(value) {
+      const memberExpressions = containers.get(value);
+      if (!memberExpressions) {
+        return value;
+      }
+      const indices = {};
+      const array = [];
+      for (const memberExpression of memberExpressions) {
+        const property = memberExpression.get('property');
+        const key = evaluatePropertyKey(property);
+        const subValue = value[key];
+        if (indices[key] == null) {
+          indices[key] = array.length;
+          array.push(subValue);
+        }
+        memberExpression.node.computed = true;
+        memberExpression.node.property = t.numericLiteral(indices[key]);
+      }
+      if (array.length === 1) {
+        for (const memberExpression of memberExpressions) {
+          const object = memberExpression.get('object');
+          memberExpression.replaceWith(object);
+        }
+        return simplifyContainers(array[0]);
+      }
+      return array.map(simplifyContainers);
+    }
+    const simplified = JSON.parse(
+      JSON.stringify(replacement, (_, value) => {
+        return simplifyContainers(value);
+      })
+    );
+    init.replaceWith(t.valueToNode(simplified));
+  }
+
+  /**
+   * Prevents collisions with outer scope
+   * @param {babel.NodePath<babel.types.Program>} path
+   * @param {Set<string>|null} taken
+   * @param {string} id
+   */
+  function rescope(path, taken, id) {
+    if (!taken) {
+      return;
+    }
+    let renamed = id;
+    while (taken.has(renamed)) {
+      renamed = path.scope.generateUid(id);
+    }
+    if (renamed === id) {
+      return;
+    }
+    path.scope.rename(id, renamed);
+    if (id === name) {
+      name = renamed;
+    }
+  }
+
   /** @type {babel.PluginObj} */
   const plugin = {
```

Comment:
Would it be easier to just generate our own JSON validator with an emphasis on code size?

Reply:
I [looked into this for a while](https://github.com/alanorozco/json-schema) and most I can say is that the spec is quite large. I got about 80% through the spec's test cases, and the remaining 20% is quite a bit of work. Not to say that drafts are added every ~2 years most of the time with non-trivial new features. 

In terms of results, my generator outputs code about half the size of ajv's default output. But the composable function syntax it outputs is also significantly slower on runtime than ajv, so the tradeoff becomes fuzzier.

Reply:
Do we need the full schema, though? If the 80% you've validated is enough, can we just roll with that?